libafl/src/mutators/string.rs

```diff
@@ -0,0 +1,533 @@
+    let selected = subcategories[selected_idx];
+
+    #[cfg(test)]
+    println!("subcategory: {} ({:?})", names[selected_idx], selected);
+
+    Some((
+        find_range(bytes, c, idx, |c| {
+            let expanded = c as u32;
+            selected.0 <= expanded && expanded <= selected.1
+        }),
+        selected,
+    ))
+}
+
+fn mutate_range<S: HasRand + HasMaxSize, F: Fn(&mut S) -> char>(
+    state: &mut S,
+    input: &mut BytesInput,
+    range: Range<usize>,
+    char_gen: F,
+) -> Result<MutationResult, Error> {
+    let temp_range = rand_range(state, range.end - range.start, MAX_CHARS);
+    let range = (range.start + temp_range.start)..(range.start + temp_range.end);
+
+    #[cfg(test)]
+    println!(
+        "{:?} => {:?}",
+        range,
+        core::str::from_utf8(&input.bytes()[range.clone()])
+    );
+
+    let replace_len = state.rand_mut().below(MAX_CHARS as u64) as usize;
+    let orig_len = range.end - range.start;
+    if input.len() - orig_len + replace_len > state.max_size() {
+        return Ok(MutationResult::Skipped);
+    }
+
+    let mut replacement = Vec::with_capacity(replace_len);
+    let mut dest = [0u8; 4];
+
+    loop {
+        let new_c = char_gen(state);
+        if replacement.len() + new_c.len_utf8() > replace_len {
+            break;
+        }
+        new_c.encode_utf8(&mut dest);
+        replacement.extend_from_slice(&dest[..new_c.len_utf8()]);
+        if replacement.len() + new_c.len_utf8() == replace_len {
+            break; // nailed it
+        }
+    }
+
+    input.bytes_mut().splice(range, replacement);
+
+    return Ok(MutationResult::Mutated);
+}
+
+impl<S> Mutator<BytesInput, S> for StringCategoryPreservingMutator
+where
+    S: HasRand + HasMaxSize,
+{
+    fn mutate(
+        &mut self,
+        state: &mut S,
+        input: &mut BytesInput,
+        _stage_idx: i32,
+    ) -> Result<MutationResult, Error> {
+        let bytes = input.bytes();
+        if let Some((c, idx)) = choose_index(state.rand_mut(), bytes) {
+            if let Some((range, category)) = choose_category_range(state.rand_mut(), bytes, c, idx)
+            {
+                #[cfg(test)]
+                println!(
+                    "{:?} => {:?}",
+                    range,
+                    core::str::from_utf8(&bytes[range.clone()])
+                );
+
+                let options: u64 = category
+                    .iter()
+                    .map(|&(start, end)| end as u64 - start as u64 + 1)
+                    .sum();
+                let char_gen = |state: &mut S| loop {
+                    let mut selected = state.rand_mut().below(options);
+                    let mut subcategories = category.iter();
+                    while let Some(&(min, max)) = subcategories.next() {
+                        if let Some(next_selected) =
+                            selected.checked_sub(max as u64 - min as u64 + 1)
```

Comment:
I don't get this part...
isn't max - min + 1 always true?

Reply:
ah never mind it's selected - subcategory_len

Reply:
category is sorted right?
I guess we can do binary search here too

Reply:
Basically, this selects uniform randomly over all entries. We use `max - min + 1` because these are inclusive ranges, and we are iterating through until we find the index that we selected uniformly.

Reply:
I don't think this is the best way to do this anymore -- I have a different strategy that I'm working on. Check in at 1400 :joy: